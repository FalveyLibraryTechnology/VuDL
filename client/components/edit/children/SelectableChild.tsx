import React, { useState } from "react";
import { useEditorContext } from "../../../context/EditorContext";
import ChildList from "./ChildList";
import AddBox from "@mui/icons-material/AddBox";
import IndeterminateCheckBox from "@mui/icons-material/IndeterminateCheckBox";
import { extractFirstMetadataValue } from "../../../util/metadata";

export interface SelectableChildProps {
    pid: string;
    selectCallback: (pid: string) => void;
    initialTitle: string;
}

export const SelectableChild = ({ pid, selectCallback, initialTitle }: SelectableChildProps): React.ReactElement => {
    const {
        state: { objectDetailsStorage },
    } = useEditorContext();
    const [expanded, setExpanded] = useState<boolean>(false);
    const loaded = Object.prototype.hasOwnProperty.call(objectDetailsStorage, pid);
    const details = loaded ? objectDetailsStorage[pid] : {};

    const title = !loaded ? initialTitle : extractFirstMetadataValue(details?.metadata ?? {}, "dc:title", "-");
    const expandControl = (
        <span onClick={() => setExpanded(!expanded)}>
            {expanded ? <IndeterminateCheckBox titleAccess="Collapse Tree" /> : <AddBox titleAccess="Expand Tree" />}
        </span>
    );
    const childList = expanded ? <ChildList pid={pid} selectCallback={selectCallback} pageSize={10} /> : "";
    return (
        <>
            {expandControl}
            <button
                onClick={() => {
                    selectCallback(pid);
                }}
            >
                {(title.length > 0 ? title : "-") + " [" + pid + "]"}
            </button>
            {childList}
        </>
    );
};

export default SelectableChild;
