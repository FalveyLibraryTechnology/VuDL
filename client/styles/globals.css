:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
