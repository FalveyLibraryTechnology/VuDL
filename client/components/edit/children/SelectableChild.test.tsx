import React from "react";
import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { SelectableChildProps, SelectableChild } from "./SelectableChild";

jest.mock("./ChildList", () => () => "ChildList");
jest.mock("../ObjectThumbnail", () => () => "ObjectThumbnail");
const mockUseEditorContext = jest.fn();
jest.mock("../../../context/EditorContext", () => ({
    useEditorContext: () => {
        return mockUseEditorContext();
    },
}));

describe("SelectableChild", () => {
    let editorValues;
    let props: SelectableChildProps;

    beforeEach(() => {
        props = { pid: "foo:123", initialTitle: "initial title", selectCallback: jest.fn() };
        editorValues = {
            state: { objectDetailsStorage: {} },
        };
        mockUseEditorContext.mockReturnValue(editorValues);
    });

    it("renders using provided default data", async () => {
        const wrapper = shallow(<SelectableChild {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders thumbnails", async () => {
        props.thumbnail = true;
        const wrapper = shallow(<SelectableChild {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders using object details storage data", async () => {
        editorValues.state.objectDetailsStorage["foo:123"] = { metadata: { "dc:title": ["loaded title"] } };
        const wrapper = shallow(<SelectableChild {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("handles missing object details storage title metadata correctly", async () => {
        editorValues.state.objectDetailsStorage["foo:123"] = { metadata: {} };
        const wrapper = shallow(<SelectableChild {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("handles empty initial titles appropriately", async () => {
        props.initialTitle = "";
        const wrapper = shallow(<SelectableChild {...props} />);
        wrapper.update();
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("allows pids to be selected", async () => {
        const wrapper = mount(<SelectableChild {...props} />);
        wrapper.find("button").simulate("click");
        expect(props.selectCallback).toHaveBeenCalledWith("foo:123");
    });

    it("can be expanded to show children", async () => {
        const wrapper = mount(<SelectableChild {...props} />);
        const expandIcon = wrapper.find("svg title");
        expect(expandIcon.text()).toEqual("Expand Tree");
        expandIcon.simulate("click");
        wrapper.update();
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
