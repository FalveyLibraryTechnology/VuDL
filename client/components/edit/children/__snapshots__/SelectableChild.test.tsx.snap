// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectableChild can be expanded to show children 1`] = `
<SelectableChild
  initialTitle="initial title"
  pid="foo:123"
  selectCallback={[MockFunction]}
>
  <span
    onClick={[Function]}
  >
    <ForwardRef(IndeterminateCheckBoxIcon)
      titleAccess="Collapse Tree"
    >
      <ForwardRef(SvgIcon)
        data-testid="IndeterminateCheckBoxIcon"
        titleAccess="Collapse Tree"
      >
        <MuiSvgIconRoot
          as="svg"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium"
          data-testid="IndeterminateCheckBoxIcon"
          focusable="false"
          ownerState={
            {
              "children": <path
                d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10H7v-2h10v2z"
              />,
              "color": "inherit",
              "component": "svg",
              "data-testid": "IndeterminateCheckBoxIcon",
              "fontSize": "medium",
              "inheritViewBox": false,
              "instanceFontSize": undefined,
              "titleAccess": "Collapse Tree",
              "viewBox": "0 0 24 24",
            }
          }
          role="img"
          viewBox="0 0 24 24"
        >
          <Insertion
            cache={
              {
                "insert": [Function],
                "inserted": {
                  "i4bv87-MuiSvgIcon-root": true,
                },
                "key": "css",
                "nonce": undefined,
                "registered": {},
                "sheet": StyleSheet {
                  "_alreadyInsertedOrderInsensitiveRule": true,
                  "_insertTag": [Function],
                  "before": null,
                  "container": <head>
                    <style
                      data-emotion="css"
                      data-s=""
                    >
                      
                      .css-i4bv87-MuiSvgIcon-root{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;width:1em;height:1em;display:inline-block;fill:currentColor;-webkit-flex-shrink:0;-ms-flex-negative:0;flex-shrink:0;-webkit-transition:fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;transition:fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;font-size:1.5rem;}
                    </style>
                  </head>,
                  "ctr": 1,
                  "insertionPoint": undefined,
                  "isSpeedy": false,
                  "key": "css",
                  "nonce": undefined,
                  "prepend": undefined,
                  "tags": [
                    <style
                      data-emotion="css"
                      data-s=""
                    >
                      
                      .css-i4bv87-MuiSvgIcon-root{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;width:1em;height:1em;display:inline-block;fill:currentColor;-webkit-flex-shrink:0;-ms-flex-negative:0;flex-shrink:0;-webkit-transition:fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;transition:fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;font-size:1.5rem;}
                    </style>,
                  ],
                },
              }
            }
            isStringTag={true}
            serialized={
              {
                "map": undefined,
                "name": "i4bv87-MuiSvgIcon-root",
                "next": undefined,
                "styles": "label:MuiSvgIcon-root;user-select:none;width:1em;height:1em;display:inline-block;fill:currentColor;flex-shrink:0;transition:fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;font-size:1.5rem;",
                "toString": [Function],
              }
            }
          />
          <svg
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="IndeterminateCheckBoxIcon"
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10H7v-2h10v2z"
            />
            <title>
              Collapse Tree
            </title>
          </svg>
        </MuiSvgIconRoot>
      </ForwardRef(SvgIcon)>
    </ForwardRef(IndeterminateCheckBoxIcon)>
  </span>
  <button
    onClick={[Function]}
  >
    initial title [foo:123]
  </button>
  <Component
    forceThumbs={false}
    pageSize={10}
    pid="foo:123"
    selectCallback={[MockFunction]}
  >
    ChildList
  </Component>
</SelectableChild>
`;

exports[`SelectableChild handles empty initial titles appropriately 1`] = `
<Fragment>
  <span
    onClick={[Function]}
  >
    <Memo(ForwardRef(AddBoxIcon))
      titleAccess="Expand Tree"
    />
  </span>
  <button
    onClick={[Function]}
  >
    - [foo:123]
  </button>
</Fragment>
`;

exports[`SelectableChild handles missing object details storage title metadata correctly 1`] = `
<Fragment>
  <span
    onClick={[Function]}
  >
    <Memo(ForwardRef(AddBoxIcon))
      titleAccess="Expand Tree"
    />
  </span>
  <button
    onClick={[Function]}
  >
    - [foo:123]
  </button>
</Fragment>
`;

exports[`SelectableChild renders thumbnails 1`] = `
<Fragment>
  <span
    onClick={[Function]}
  >
    <Memo(ForwardRef(AddBoxIcon))
      titleAccess="Expand Tree"
    />
  </span>
  <button
    onClick={[Function]}
  >
    initial title [foo:123]
    <br />
    <Component
      pid="foo:123"
    />
  </button>
</Fragment>
`;

exports[`SelectableChild renders using object details storage data 1`] = `
<Fragment>
  <span
    onClick={[Function]}
  >
    <Memo(ForwardRef(AddBoxIcon))
      titleAccess="Expand Tree"
    />
  </span>
  <button
    onClick={[Function]}
  >
    loaded title [foo:123]
  </button>
</Fragment>
`;

exports[`SelectableChild renders using provided default data 1`] = `
<Fragment>
  <span
    onClick={[Function]}
  >
    <Memo(ForwardRef(AddBoxIcon))
      titleAccess="Expand Tree"
    />
  </span>
  <button
    onClick={[Function]}
  >
    initial title [foo:123]
  </button>
</Fragment>
`;
